src/NuGet.Core/NuGet.Commands/RestoreCommand/SourceRepositoryDependencyProvider.cs

```diff
@@ -232,5 +284,86 @@ private IEnumerable<LibraryDependency> GetDependencies(FindPackageByIdDependency
                 }
             }
         }
+
+        /// <summary>
+        /// Discover all package versions from a feed.
+        /// </summary>
+        private async Task<IEnumerable<NuGetVersion>> GetPackageVersionsAsync(string id,
+                                                                    SourceCacheContext cacheContext,
+                                                                    ILogger logger,
+                                                                    CancellationToken cancellationToken)
+        {
+            IEnumerable<NuGetVersion> packageVersions = null;
+            try
+            {
+                if (_throttle != null)
+                {
+                    await _throttle.WaitAsync();
+                }
+                packageVersions = await _findPackagesByIdResource.GetAllVersionsAsync(
+                    id,
+                    cacheContext,
+                    logger,
+                    cancellationToken);
+            }
+            catch (FatalProtocolException e) when (_ignoreFailedSources)
+            {
+                if (!_ignoreWarning)
+                {
+                    _logger.LogWarning(e.Message);
+                }
+                return null;
+            }
+            finally
+            {
+                _throttle?.Release();
+            }
+
+            return packageVersions;
+        }
+
+        /// <summary>
+        /// Cache on LibraryRange and TargetFramework
+        /// </summary>
+        private class DependencyInfoCacheKey : IEquatable<DependencyInfoCacheKey>
```

Comment:
I found the same cache key class existed under LibraryRangeCacheKey and switched to that, removing this private class.

I would normally agree with simplifying here, but this path in the walk is critical for perf. Every package in the unflattened graph needs to be resolved, which is much larger than the number you see in the final flattened graph in the assets file.